SAMZA-2638: Skip container stop/restart for processors that have same work assignment across rebalances

mynameborat opened the pull request:
**Problem**:
As part of rebalance, we always expire the current work assignment and proceed to signal consensus and start the container with new work assignment. It is inefficient for the processors to do the above when there is no changes in the work assignment between the active job model & the proposed job model.

**Changes**:
 - Processors perform onJobModelExpired and onNewJobModel only if there are changes to their work assignment across the active and proposed job model.
 - Processors no longer shutdown if they are not part of job model. Refer to the [SAMZA-2638](https://issues.apache.org/jira/browse/SAMZA-2638) for explanations
 - Add helper methods in `JobModelUtil` for work assignment comparison

**Tests**:
 - Unit tests for job model expiration
 - Unit tests for on new job model
 - Unit tests for job model version change flow
 - Unit tests for container model comparison

**API Changes**: None

**Usage Instructions**: None

**Upgrade Instructions**: None

mynameborat commented:
The test failures are unrelated to the change and are happening on the master branch. @atoomula is looking into them.
@shanthoosh @sborya can you guys help with the reviews?

sborya commented on samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java:
```diff
@@ -392,6 +395,78 @@ StartpointManager createStartpointManager() {
+    if (JobModelUtil.compareContainerModelForProcessor(processorId, activeJobModel, newJobModel)) {
+      LOG.info("Skipping job model expiration for processor {} due to no change in work assignment.", processorId);
+    } else {
+      coordinatorListener.onJobModelExpired();
```
add a LOG line here too.

sborya commented on samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java:
```diff
@@ -392,6 +395,78 @@ StartpointManager createStartpointManager() {
+    Preconditions.checkNotNull(newJobModel, "JobModel cannot be null. Failing onNewJobModel");
+    // start the container with the new model
+    if (!JobModelUtil.compareContainerModelForProcessor(processorId, activeJobModel, newJobModel)) {
+      if (newJobModel.getContainers().containsKey(processorId)) {
```
what happens in the else{ } case?

mynameborat replied:
The implication of work assignment remaining the same can be categorized into 
1. Processor part of the job model
2. Processor not part of the job model.

For both the state of the processor remains what it was when the rebalance started. e.g., 
[1] should continue to process its work assignment without any interruption as part of the rebalance. i.e., there will be no expiration of the existing work (a.k.a samza container won't be stopped) and also no notification to `StreamProcessor` about the rebalance since work assignment didn't change.
[2] should have no work and be idle processor and will continue to be idle.

shanthoosh commented on samza-core/src/main/java/org/apache/samza/job/model/JobModelUtil.java:
```diff
@@ -119,6 +120,47 @@ public static JobModel readJobModel(String jobModelVersion, MetadataStore metada
+      return false;
+    }
+
+    return compareContainerModel(first.getContainers().get(processorId), second.getContainers().get(processorId));
```
Do you think it will be useful to add NPE guards against the result of the object obtained by `jobModel.getContainers()`

mynameborat replied:
JobModel guarantees `getContainers()` always returns non-null map instance.

shanthoosh commented on samza-core/src/main/java/org/apache/samza/job/model/JobModelUtil.java:
```diff
@@ -119,6 +120,47 @@ public static JobModel readJobModel(String jobModelVersion, MetadataStore metada
+   *         false - otherwise
+   */
+  @VisibleForTesting
+  static boolean compareContainerModel(ContainerModel first, ContainerModel second) {
```
Minor:
I may be wrong. But AFAIK, I think `Objects.equals` from java-native already has all these NPE checks . If possible, then we can re-use them.

shanthoosh commented on samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java:
```diff
@@ -392,6 +395,90 @@ StartpointManager createStartpointManager() {
+    }
+
+    if (JobModelUtil.compareContainerModelForProcessor(processorId, activeJobModel, newJobModel)) {
+      LOG.info("Skipping job model expiration for processor {} due to no change in work assignment.", processorId);
```
Minor:
Would be helpful to either log the jobModel or the work-assignment here for debugging purposes. Essentially logging the before/after work-assignment on a quorum change just for debugging purposes. May-be logging it when they're different would be better.

shanthoosh commented:
Majority of the above comments/suggestion are minor. We can may-be address them in a follow-up PR if there's agreement on their utility.

mynameborat commented:
> Majority of the above comments/suggestion are minor. We can may-be address them in a follow-up PR if there's agreement on their utility.

Addressed the comments as part of #1484
